src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -1587,5 +1603,73 @@ protected override void Dispose (bool disposing)
 				stream?.Dispose ();
 			}
 		}
+
+#if NET
+		static class CertificateHelper {
+			// Based on https://github.com/dotnet/runtime/blob/c2848c582f5d6ae42c89f5bfe0818687ab3345f0/src/libraries/Common/src/System/Net/Security/CertificateHelper.cs
+			// with the NetEventSource code removed and namespace changed.
+			const string ClientAuthenticationOID = "1.3.6.1.5.5.7.3.2";
+
+			internal static X509Certificate2? GetEligibleClientCertificate (X509CertificateCollection? candidateCerts)
+			{
+				if (candidateCerts is null || candidateCerts.Count == 0) {
+					return null;
+				}
+
+				var certs = new X509Certificate2Collection ();
+				certs.AddRange (candidateCerts);
+
+				return GetEligibleClientCertificate (certs);
+			}
+
+			internal static X509Certificate2? GetEligibleClientCertificate (X509Certificate2Collection? candidateCerts)
+			{
+				if (candidateCerts is null || candidateCerts.Count == 0) {
+					return null;
+				}
+
+				foreach (X509Certificate2 cert in candidateCerts) {
+					if (!cert.HasPrivateKey) {
+						continue;
+					}
+
+					if (IsValidClientCertificate (cert)) {
+						return cert;
+					}
+				}
+				return null;
+			}
+
+			static bool IsValidClientCertificate (X509Certificate2 cert)
+			{
+				foreach (X509Extension extension in cert.Extensions) {
+					if ((extension is X509EnhancedKeyUsageExtension eku) && !IsValidForClientAuthenticationEKU (eku)) {
+						return false;
+					} else if ((extension is X509KeyUsageExtension ku) && !IsValidForDigitalSignatureUsage (ku)) {
+						return false;
+					}
+				}
+
+				return true;
+			}
+
+			static bool IsValidForClientAuthenticationEKU (X509EnhancedKeyUsageExtension eku)
+			{
+				foreach (System.Security.Cryptography.Oid oid in eku.EnhancedKeyUsages) {
+					if (oid.Value == ClientAuthenticationOID) {
+						return true;
+					}
+				}
+
+				return false;
+			}
+
+			static bool IsValidForDigitalSignatureUsage (X509KeyUsageExtension ku)
+			{
+				const X509KeyUsageFlags RequiredUsages = X509KeyUsageFlags.DigitalSignature;
+				return (ku.KeyUsages & RequiredUsages) == RequiredUsages;
```

Comment:
I have always preferred the [HasFlag](https://learn.microsoft.com/en-us/dotnet/api/system.enum.hasflag?view=net-8.0) method over the &, but is a matter of taste. I just think it is easier to read. 
```suggestion
				return (ku.KeyUsages.HasFlag(X509KeyUsageFlags.DigitalSignature);
```

Reply:
Since this is code taken from the .NET Runtime I tried not changing the logic from its origin.